src/python/pants/backend/codegen/tasks/wire_gen.py

```diff
@@ -74,11 +77,53 @@ def resolve_deps(self, unresolved_deps):
   def synthetic_target_extra_dependencies(self, target):
     return self.resolve_deps(self.get_options().javadeps)
 
+  def _collect_unpacked_sources(self, target):
+    unpacked_sources = self.context.products.get_data('unpacked_archives', dict)
+    for dependency in target.dependencies:
+      if isinstance(dependency, UnpackedJars):
+        found_files, rel_unpack_dir = unpacked_sources[dependency]
+        yield (found_files, rel_unpack_dir)
+
+  def _collect_sources_in_source_root(self, target):
+    source_root = SourceRoot.find(target)
+    abs_source_root = os.path.join(get_buildroot(), source_root)
+    for dirpath, dirnames, filenames in os.walk(abs_source_root):
+      proto_files = [filename for filename in filenames if filename.lower().endswith('.proto')]
+      proto_files = [os.path.join(os.path.relpath(dirpath, abs_source_root), filename)
+                     for filename in proto_files]
+      if proto_files:
+        yield (proto_files, source_root)
+
   def execute_codegen(self, targets):
     # Invoke the generator once per target.  Because the wire compiler has flags that try to reduce
     # the amount of code emitted, Invoking them all together will break if one target specifies a
     # service_writer and another does not, or if one specifies roots and another does not.
+
+    symlinks = []
+    def create_symlink_if_not_present(source, link_name):
+      if source == link_name:
+        return
+      if os.path.exists(source) and not os.path.exists(link_name):
+        if not os.path.exists(os.path.dirname(link_name)):
+          os.makedirs(os.path.dirname(link_name))
+        os.symlink(source, link_name)
+        symlinks.append(link_name)
+
     for target in targets:
+      protos_repo = os.path.join(self.workdir, 'protos-repo')
```

Comment:
I am not sure that it is valid to create a single dir for all symlinks ('.pants.d/wire_gen/protos-repo')   If that's what's happening, we just re-introduced a global dir into our isolated gen strategy.  To stay isolated, we would have to re-create this symlink dir of dependent wire_proto sources separately for each target we're compiling.

 I'd like to get wire to support multiple instances of --proto_path.  If we can get that in place, all this code gets deleted.  Can we lift all of this out code dealing with `protos_repo` into another function and isolate it some more?